sdk/core/Azure.Core/src/HttpMessage.cs

```diff
@@ -14,12 +14,8 @@ namespace Azure.Core
     /// </summary>
     public sealed class HttpMessage : IDisposable
     {
-        /// <summary>
-        /// This dictionary is keyed with <c>Type</c> for a couple of reasons. Primarily, it allows values to be stored such that even if the accessor methods
-        /// become public, storing values keyed by internal types make them inaccessible to other assemblies. This protects internal values from being overwritten
-        /// by external code. See the <see cref="TelemetryDetails"/> and <see cref="UserAgentValueKey"/> types for an example of this usage.
-        /// </summary>
-        private Dictionary<Type, object>? _typeProperties;
+        private ArrayBackedPropertyBag<string, object>? _properties;
+        private readonly ArrayBackedPropertyBag<long, object> _propertyBag = new();
```

Comment:
The tradeoff is between simplifying the logic and optimizing for the type based scenario rather than saving the size of the field reference in the allocation of `HttpMessage`. I opted to go for the simplification, but I'm open to going the other way if you feel strongly that it's better.

Reply:
Maybe I miss something, but can't we use something like this as a key:
```cs
public readonly struct Key : IEquatable<Key>
{
    private readonly object _key;

    public Key(Type type)
    {
        _key = type;
    }

    public Key(string s)
    {
        _key = s;
    }

    public override bool Equals(object obj) => obj is Key other && Equals(other);

    public bool Equals(Key other) =>
        _key is Type type && other._key is Type otherType 
            ? type == otherType 
            : _key is string s && other._key is string otherS && s == otherS;

    public override int GetHashCode() => _key.GetHashCode();
}
```

Reply:
If `ArrayBackedPropertyBag` becomes a struct, then accessing it as one of the values will cause boxing/unboxing.

Reply:
@AlexanderSher, yes, but I am not sure we care about this code path. It's basically legacy. Also, if we make the bag be a struct, I would keep the long-object one as a Dictionary. The new bag is very large (8 longs), even when empty, and so we don't want it to be inlined in the message. Dictionary when empty is just one null long.